### PR TITLE
Add internal unsubscription if we know the websocket has disconnected

### LINKED ISSUE #86: Unsubscribe event listener when websocket client disconnects.
We should internally unsubscribe from the event subject when the websocket disconnects. 
Disconnect event calls https://github.com/trustlines-network/relay/blob/develop/relay/api/streams/transport.py#L45

### PR DESCRIPTION
**Pull request by berndbohmeier**
The client keeps a list of subscriptions now so that we can stop listening if the connection is closed
closes #86

### CODE REVIEW

File: relay/streams.py [hunk 1]
```diff
@@ -12,21 +12,63 @@
...
+    def register(self, subscription: 'Subscription') -> None:
+        """
+        Registers a subscription that this client has done.
+        On closing the connection with `close` these subscription will get unsubscribed
+        """
+        self.subscriptions.append(subscription)
+
+    def unregister(self, subscription: 'Subscription') -> None:
+        """
+        Unregisters a subscription that this client is not listing for anymore.
```
Comment (schmir) [inline relay/streams.py]: listing -> listening

File: relay/streams.py [hunk 2]
```diff
@@ -12,21 +12,63 @@
...
+            DisconnectedError: This is raised if the client has already disconnected.
+        """
         raise NotImplementedError
 
+    def close(self) -> None:
+        """
+        Should be called when the connection to this client is closed.
+        This will unsubscribe all registered subscriptions
+        """
+        for subscription in self.subscriptions:
```
Comment (schmir) [inline relay/streams.py]: This looks like we may have to iterate over a copy of self.subscriptions. Could that be the case?

Comment (berndbohmeier) [reply relay/streams.py]: I dont think so, as we dont delete from that list while iterating. I can still do it to be sure. 
But I think i should probably delete the subscriptions at the end, which I dont do.

Comment (schmir) [reply relay/streams.py]: Subscription.unsubscribe calls client.unregister, unregister seems to call self.subscriptions.remove(subscription), so this looks like it is modifying that list. Could also be the case that it's another list. I'm not sure either, what is being modified here.

Comment (berndbohmeier) [reply relay/streams.py]: Those are two different lists. A subjects keeps track of its subscription, and a client keeps track where he subscribed to

Comment (berndbohmeier) [reply relay/streams.py]: I will copy it to be sure.

Comment (berndbohmeier) [reply relay/streams.py]: No wait, actually you are right. My bad.

**Comment by schmir** [comment]
two of the three new test cases seem unrelated, at least they don't fail if the list is not being copied.

Please split such commits next time and add a bit more information.  The reviewer will thank you.

If we fix bugs and write tests for them, I would really like us to write a failing test,commit that, and commit the fix afterwards. Not sure if that works with our CI setup.

I feel like streams.py could/should be simplified. If it was simple enough we wouldn't have the discussions if the list that is being iterrated over, is also being modified or not. But that's probably a different issue.

Nevertheless please push this as is.

**Pull request merged**